src/lib/cache/helper.go

```diff
@@ -0,0 +1,88 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/goharbor/harbor/src/lib/log"
+)
+
+var (
+	onceMu = keyMutex{m: &sync.Map{}}
+)
+
+// Once retrieve the cached key value, build and cache the value when the cached key value not found
+func Once(c Cache, key string, value interface{}, builder func() (interface{}, error), expiration ...time.Duration) error {
```

Comment:
The `Once` func is confusing, you call it more than once... also, having to add the first parm `c` means it should probably be moved as c's attribute.

It'd be better moved to the cache.Cache, introduce something like `loader` if the cache misses a key it always try to load it from the real source.  But then you need to introduce different loader for the cache b/c different key may map to different source.

Another simplified option I think is to move it to the caller(storage/driver) layer, so cache user need to worry about initializing the cache, if it misses the caller write the data to the cache.

Reply:
Rename this function name to `FetchOrSave`.

It's not very well to move this function to the `Cache` interface because this is a pattern to how to use cache to improve the performance of an expensive operation. This pattern is that try to fetch the value from the cache,  compute the value save it to the cache when it's not found in the cache so that next time we can fetch it from the cache directly.

This function does one more thing than the pattern is that it uses a lock to avoid computing and saving the value multi times in high concurrency.

The `intializer` is not part of the cache, it should be work together with a key. For example, we may only have one default Redis cache in the system. This cache can be used in the config manager, project DAO. I don't think `intializer` for the cache can match these scenarios.